src/api/conceptApi.ts

```diff
@@ -35,3 +35,32 @@ export async function searchConcepts(
     metaImage: res.metaImage,
   }));
 }
+
+export async function fetchConcepts(
+  conceptIds: string[],
+  context: Context,
+): Promise<GQLConcept[]> {
+  const response = await Promise.all(
+    conceptIds.map(id => fetch(`/concept-api/v1/concepts/${id}`, context)),
+  );
+  const concepts = await Promise.all(
+    response.map(async concept => {
+      try {
+        return await resolveJson(concept);
+      } catch (e) {
+        return undefined;
+      }
+    }),
+  );
+  return concepts.reduce((acc: GQLConcept[], res: SearchResultJson) => {
+    if (res !== undefined) {
+      acc.push({
+        id: res.id,
+        title: res.title.title,
+        content: res.content.content,
+        metaImage: res.metaImage,
+      });
+    }
+    return acc;
+  }, []);
```

Comment:
Ville egentlig bruke reduce på den første mappingen som under, for å slippe å returnere og sjekke for undefined etterpå. Men synes det ble rotete og mye unødvendige awaits på previousPromise osv. Men tar gjerne innspill på dette!
```
const concepts = await response.reduce(async (previousPromise, concept) => {
    let acc = await previousPromise;
    try {
      const res = await resolveJson(concept);
      acc.push(res);
      return acc;
    } catch (e) { return acc; }
  }, Promise.resolve([]));
  return concepts.map((res: SearchResultJson) => ({
    id: res.id,
    title: res.title.title,
    content: res.content.content,
    metaImage: res.metaImage,
  }));
```

Reply:
trenger det å være noen `reduce` i det hele tatt?
Troor jeg syns en `map` er penere i dette tilfellet
(Må innrømme at jeg har ikke heeelt koll på hva som vil ha best ytelse når alt er async da)

```typescript
export async function fetchConcepts(
  conceptIds: string[],
  context: Context,
): Promise<GQLConcept[]> {
  return (await Promise.all(conceptIds.map(async id => {
      const concept = await fetch(`/concept-api/v1/concepts/${id}`, context);
      try{
        const res: SearchResultJson = await resolveJson(concept);
        const result: GQLConcept = {
          id: res.id,
          title: res.title.title,
          content: res.content.content,
          metaImage: res.metaImage,
        }
        return result;
      } catch(e) {
        return undefined;
      }
  }))).filter(c => !!c);
}
```

Spesielt om vi ikke hånterer feil api-respons her. Ser ikke at vi gjør det i andre klienter?

```typescript
export async function fetchConcepts(
  conceptIds: string[],
  context: Context,
): Promise<GQLConcept[]> {
  return Promise.all(
    conceptIds.map(async id => {
      const concept = await fetch(`/concept-api/v1/concepts/${id}`, context);
      const res: SearchResultJson = await resolveJson(concept);
      const result: GQLConcept = {
        id: res.id,
        title: res.title.title,
        content: res.content.content,
        metaImage: res.metaImage,
      };
      return result;
    }),
  );
}
```

Reply:
Hm ja, det ble en del bedre. Ble litt opphengt i å slippe filtrering på undefined, men er jo like greit å filtrere når jeg uansett må iterere over lista to ganger. Takk! Må håndtere feil api-respons fordi man kan legge til forklaringer som ikke er publisert som får 404 på fetch, og da ryker hele Promise.all.

Reply:
Ah riktig.
Du kan jo slippe å iterere over lista 2 ganger med en reduce med noe som ligner veldig på det du hadde, men tror den er litt lite effektiv siden vi må awaite forrige concept i hver iterasjon for å kunne appende til arrayen?
```typescript
export async function fetchConcepts(
  conceptIds: string[],
  context: Context,
): Promise<GQLConcept[]> {
  return conceptIds.reduce(
    async (accConcepts: Promise<GQLConcept[]>, conceptId: string) => {
      const resp = await fetch(`/concept-api/v1/concepts/${conceptId}`, context);
      try {
        const parsed: SearchResultJson | undefined = await resolveJson(resp);
        if (parsed !== undefined) {
          const acc = await accConcepts;
          acc.push({
            id: parsed.id,
            title: parsed.title.title,
            content: parsed.content.content,
            metaImage: parsed.metaImage,
          });
          return acc;
        }
      } catch (e) {
        return accConcepts;
      }
      return accConcepts;
    },
    Promise.resolve([]),
  );
}
```

Den "optimale" løsningen er vel med mutability på en eller annen måte, kanskje noe sånt?
Vet ikke om man må awaite promisene fra map'en når outputen ikke brukes :zany_face: 
```typescript
export async function fetchConcepts(
  conceptIds: string[],
  context: Context,
): Promise<GQLConcept[]> {
  const acc: GQLConcept[] = [];
  await Promise.all(conceptIds.map(async (conceptId: string) => {
    const resp = await fetch(`/concept-api/v1/concepts/${conceptId}`, context);
    try {
      const parsed: SearchResultJson | undefined = await resolveJson(resp);
      if (parsed !== undefined) {
        acc.push({
          id: parsed.id,
          title: parsed.title.title,
          content: parsed.content.content,
          metaImage: parsed.metaImage,
        });
      }
    } catch (e) {
      return Promise.resolve();
    }
  }));
  return acc;
}
```

Meeen jeg tipper map-løsningen er effektiv nok (og den er penest imo) siden lista vil sikkert aldri være lang nok til at det blir significant å iterere over den flere ganger (I alle fall ikke sammenlignet med nettverkskallene for å hente forklaringene).